vllm/model_executor/models/eagle.py

```diff
@@ -0,0 +1,86 @@
+from typing import Iterable, List, Optional, Tuple
+
+import torch
+import torch.nn as nn
+
+from vllm.attention.backends.abstract import AttentionMetadata
+from vllm.model_executor.model_loader.weight_utils import default_weight_loader
+from vllm.model_executor.models import ModelRegistry
+from vllm.model_executor.sampling_metadata import SamplingMetadata
+from vllm.sequence import IntermediateTensors, SamplerOutput
+from vllm.transformers_utils.configs.eagle import EAGLEConfig
+
+
+class EAGLE(nn.Module):
+
+    def __init__(self, config: EAGLEConfig, *args, **kwargs) -> None:
+        super().__init__()
+        self.config = config
+
+        architectures = getattr(self.config.model, "architectures", [])
+        for arch in architectures:
+            model_cls = ModelRegistry.load_model_cls(arch)
+            if model_cls is not None:
+                break
+
+        self.model = model_cls(self.config.model, *args, **kwargs)
+        self.fc = nn.Linear(config.model.hidden_size * 2,
+                            config.model.hidden_size,
+                            bias=False)
+
+        self.token_map = None
```

Comment:
This is used for vocab truncation (as introduced in Medusa as well). The code wasn't complete for this, I've made the changes necessary to add complete support and have added the comments in code as well.

Reply:
Can this be disabled? If not, can you point me to the logic in the reference implementation?

Reply:
This is disabled by default and is only used if the checkpoint file of the draft model has an explicit token_map tensor and config has an optional attribute truncated_vocab_size < vocab_size. I am adding this clarification to both EAGLE and Medusa in the comments and the docstring.

This wasn't there in any of the reference implementations but we have found it to work well as it reduces overhead with minimal impact on acceptance rate. To use it, you must find the most frequent tokens in your target data distribution and add them as token_map to the draft checkpoint.

Reply:
It would be awesome to have this explanation in the docs or in a docstring. Cool technique.

Seems also the testing story is weak for this (correct me if wrong), but we can follow up later since it's not new to this PR.

Reply:
Added the explanation to both Medusa's and EAGLE's doctring.